spec.html

```diff
@@ -26403,6 +26494,169 @@ <h1>Cyclic Module Records</h1>
+              </td>
+            </tr>
+          </table>
+        </emu-table>
+
+        <emu-clause id="sec-LoadRequestedModules" type="concrete method">
+          <h1>LoadRequestedModules ( [ _hostDefined_ ] ): a Promise object</h1>
+          <dl class="header">
+            <dt>for</dt>
+            <dd>a Cyclic Module Record _module_</dd>
+
+            <dt>description</dt>
+            <dd>It populates the [[LoadedModules]] of all the Modue Records in the dependency graph of _module_ (most of the work is done by the auxiliary function InnerModuleLoading). It takes an optional _hostDefined_ parameter, that is passed to the HostLoadImportedModule hook.</dd>
+          </dl>
+
+          <emu-alg>
+            1. If _hostDefined_ is not present, let _hostDefined_ be *undefined*.
+            1. Let _pc_ be ! NewPromiseCapability(%Promise%).
+            1. Let _state_ be a new ModuleLoadState Record { [[Action]]: ~graph-loading~, [[IsLoading]]: *true*, [[PendingModules]]: 1, [[Visited]]: a new empty List, [[PromiseCapability]]: _pc_, [[HostDefined]]: _hostDefined_ }.
+            1. Perform InnerModuleLoading(_state_, _module_).
+            1. Return _pc_.[[Promise]].
+          </emu-alg>
+
+          <emu-note>
+            The _hostDefined_ parameter can be used to pass additional information necessary to fetch the imported modules. It is used, for example, by HTML to set the correct fetch destination for <code>&lt;link rel="preload" as="..."&gt;</code> tags.
+            <code>import()</code> expressions never set the _hostDefined_ parameter.
+          </emu-note>
+
+          <emu-clause id="sec-InnerModuleLoading" type="abstract operation">
+            <h1>
+              InnerModuleLoading (
+                _state_: a ModuleLoadState Record whose [[Action]] is ~graph-loading~,
+                _module_: a Module Record,
+              ): ~unused~
+            </h1>
+            <dl class="header">
+              <dt>description</dt>
+              <dd>It is used by LoadRequestedModules to recursively perform the actual loading process for _module_'s dependency graph.</dd>
+            </dl>
+
+            <emu-alg>
+              1. Assert: _state_.[[IsLoading]] is *true*.
+              1. If _module_ is a Cyclic Module Record, _module_.[[Status]] is ~new~, and _state_.[[Visited]] does not contain _module_, then
+                1. Add _module_ to _state_.[[Visited]].
+                1. Let _requestedModulesCount_ be the length of _module_.[[RequestedModules]].
+                1. Set _state_.[[PendingModules]] to _state_.[[PendingModules]] + _requestedModulesCount_.
+                1. For each String _required_ of _module_.[[RequestedModules]], do
+                  1. If _module_.[[LoadedModules]] contains a Record _record_ such that _record_.[[Specifier]] is _required_, then
+                    1. Perform ContinueModuleLoading(_state_, NormalCompletion(_record_.[[Module]])).
```

Comment:
Does `ContinueModuleLoading` really need to be a separate function? The logic seems short enough to me that it might actually be worth inlining. Even though it causes a little duplication, I believe it would also lead to some simplifications and clearer reading.

Reply:
I like it as a separate function because it allows keeping the loading logic clearly separated in three sections:
1- AOs for static imports: `LoadRequestedModules`, `InnerModuleLoading`, and `ContinueModuleLoading`
2. AOs for dynamic imports: Runtime Semantics of `import ( AssignmentExpression )` and `ContinueDynamicImport`
3. AOs used by both to interact with the host: `HostLoadImportedModule` and `FinishLoadingImportedModule`, which just caches and dispatches back to 1 or 2

Reply:
When reading the `ContinueModuleLoading` call in `LoadRequestedModules` one has to jump to `ContinueModuleLoading`, which then already passes the "isLoaded" check and also the normal completion check so that it's effectively just a one-liner. Perhaps just inlinine it here only then? It's easier to see that `LoadRequestedModules` is a recursive function that way.

Reply:
The [[IsLoading]] check is still needed, because a call to HostLoadImportedModule done previously in the loop might have set it to false. However, you made me realize that we need that check in LoadReqeustedModules anyway, to avoid calling HostLoadImportedModule when [[IsLoading]] becomes false.